blockchain/chain/src/store/tipset_tracker.rs

```diff
@@ -1,47 +1,55 @@
 // Copyright 2019-2022 ChainSafe Systems
 // SPDX-License-Identifier: Apache-2.0, MIT
 
-use log::{debug, warn};
-use std::collections::HashMap;
-
 use cid::Cid;
 use forest_blocks::{BlockHeader, Tipset};
+use forest_networks::ChainConfig;
 use forest_utils::db::BlockstoreExt;
 use fvm_ipld_blockstore::Blockstore;
 use fvm_shared::clock::ChainEpoch;
+use log::{debug, warn};
 use parking_lot::Mutex;
+use std::{collections::BTreeMap, sync::Arc};
 
 use super::Error;
 
 /// Tracks blocks by their height for the purpose of forming tipsets.
 #[derive(Default)]
 pub(crate) struct TipsetTracker<DB> {
-    entries: Mutex<HashMap<ChainEpoch, Vec<Cid>>>,
+    entries: Mutex<BTreeMap<ChainEpoch, Vec<Cid>>>,
     db: DB,
+    chain_config: Arc<ChainConfig>,
 }
 
 impl<DB: Blockstore> TipsetTracker<DB> {
-    pub fn new(db: DB) -> Self {
+    pub fn new(db: DB, chain_config: Arc<ChainConfig>) -> Self {
         Self {
             entries: Default::default(),
             db,
+            chain_config,
         }
     }
 
     /// Adds a block header to the tracker.
     pub fn add(&self, header: &BlockHeader) {
-        let cids = {
-            let mut map = self.entries.lock();
-            let cids = map.entry(header.epoch()).or_default();
-            if cids.contains(header.cid()) {
-                debug!("tried to add block to tipset tracker that was already there");
-                return;
-            }
-            cids.push(*header.cid());
-            cids.clone()
-        };
+        let mut map = self.entries.lock();
+        let cids = map.entry(header.epoch()).or_default();
+        if cids.contains(header.cid()) {
+            debug!("tried to add block to tipset tracker that was already there");
+            return;
+        }
+        self.check_multiple_blocks_from_same_miner(cids, header);
+        cids.push(*header.cid());
+        drop(map);
```

Comment:
I don't get why we can not use an implicit drop of the guard like before, using a block expression?

Reply:
We can, but I found explicit to be more readable and easier to change; no indentations needed, and the same number of LoC. If you feel strongly about it, I'm okay with falling back to blocks. What do you think?

Reply:
I'm okay with both, just a preference style for scopes on my side.